### PR TITLE
feat: publish FPM gym overview page

### PR DESCRIPTION
**Pull request by simone-chen**
## Summary

- Match FPM to the E2E accuracy UX: shared slate/indigo styles, compact AISimulate header and branch control, summary cards, table, and a light/dark preference preserved across both pages.
- Add the public FPM Gym Overview with expandable configurations, sorting, measurement counts, MAPE, coverage, pinned HF evidence, and explicit unavailable/stale states. Keep FPM (KV warmup on), FPM (KV warmup off), and Regression; exclude op-based and all other Gym tabs.
- Move served page sources and assets into root `pages/`, preserving public URLs and older E2E release snapshot reads. Keep universe unpublished and Rust design HTML at its existing documentation path.
- Add daily evaluation at 10:47 UTC for main and numeric releases >= 0.12.0, plus exact-SHA manual dispatch. Pin HF once, prepare verified exact wheels, and evaluate at most two branches concurrently.
- Store branch results in 90-day Actions artifacts. Validate campaign identity, completion, hashes, and source ancestry before Pages publishes aggregates; retain prior valid results after failures and publish successful branches independently.
- Port development-only evaluation code from Gym commit `e8221729db2802e822f6919fd68bc2941743385b` under confirmed Apache-2.0 permission. Preserve shared measurement membership, rank handling, worker isolation, predict → score → tune ordering, and coverage/MAPE/artifact-ID variant selection. Synchronize notices and generated CODEOWNERS.

## Screenshot

- Latest overview with real pipeline results: Regression first, followed by FPM with KV warmup on/off; zero-measurement configurations hidden.

FPM accuracy overview showing Regression before FPM and 15 measured configurations

## Validation

- Resolve the final review findings at `90c08444`: reject duplicate keys in HF catalogs, manifests, and sidecars; label combined source issues “excluded or unavailable.” Pass 368 focused Python tests, 21 workflow tests, Chromium checks, and real pinned-HF parsing for 20 configurations and 28 sidecars.
- Pass the [full hosted main campaign](https://github.com/ai-dynamo/aisimulate/actions/runs/35279017422): build the exact Linux wheel for `a0c3c49dd3d56b4746ca3fcf8aed17b88d6e680c`, evaluate HF `5487a4599a7fbc012c07bcd3699754bdf4a8bef7`, and upload 90-day overview artifacts. Complete all 20 configurations across six models; 15 contain 2,155,267 eligible observations and five have no measurements.
- Validate the downloaded ZIP layout, summary/qualification schemas, summary and wheel SHA-256 hashes, run identity, evaluator revision, source SHA, HF revision, and branch completion. Build Pages from the actual artifact and pass Chromium checks for real model/configuration rows, provenance, sorting, expansion, labels, and mobile layout.
- Fix two issues found by hosted runs: support HF's shared blob cache while retaining path/hash checks, and upload one output directory so the container runner preserves both JSON files at the artifact root. All 91 FPM tests and the PR checks pass with both fixes.
- Pass offline HF hash/protocol/rank tests, evaluation and publication contracts, E2E compatibility tests, repository policy tests, and 56 JavaScript UI/workflow tests.
- Pass Chromium checks for FPM branch selection, sorting, expansion, unavailable/stale states, late responses, light/dark theme persistence across both accuracy pages, keyboard theme controls, and mobile layout; existing FPE browser checks also pass.
- Pass actionlint, CI-pinned Ruff checks/formatting, documentation destinations, copyright and packaged notices, CODEOWNERS coverage/generation, and the Pages build. Confirm Rust design HTML is unchanged and universe is absent from the deployment.
- Run real HF smoke evaluations with locally built exact native wheels: main `c3cc310979295e33bec93f849df906c6b8a847bc` and release/0.12.0 `1f728534910187ecb2f021ef5e4bd4949bd555d0`, using HF revision `5487a4599a7fbc012c07bcd3699754bdf4a8bef7`. Both use the same 157,007 MiniMax-M2.7/H200/vLLM observations; warmup FPM predicts 156,888 with approximately 4.40% MAPE. Main regression predicts 150,422 with approximately 4.63% MAPE.

## Scope and limitations

- Accuracy is advisory: the full main run predicts 1,144,609 observations with KV warmup on (3.07% MAPE), 749,261 with KV warmup off (10.53%), and 2,056,083 with Regression (8.80%). Missing reviewed inputs, cold starts, and prediction errors remain in coverage. Main's native FPM rejects Kimi-K3's encoder/multimodal model; its errors remain visible rather than being dropped.
- The hosted campaign uses an isolated validation branch and the registered Pages dispatch entry point, with the same evaluator and FPM reusable workflow as this PR. Production trust checks exclude this run from publication; no Pages deployment occurred. Scheduled main dispatch and production artifact discovery remain post-merge checks. The release smoke is partial and remains separate from the full hosted main campaign and synthetic fixture tests.
- The tested release/0.12.0 wheel lacks Gym's worker-scoped regression API. Display Regression as unsupported and retain all measurements in its coverage denominator; do not substitute a different regression contract. This smoke configuration has no reviewed no-warmup input, which is also explicitly unavailable.
- Do not change the HF dataset, existing GitLab deployment, E2E evaluation policy, or release staging gates. High MAPE remains advisory. Before the first completed campaign, Pages shows “No completed evaluation.”

### CODE REVIEW

File: scripts/fpm_accuracy/types/worker_config.py [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+
+    if (model_kind or "").strip().lower() != "moe":
+        return None, None
+    width = tp_size or 1
+    if expert_parallel_enabled:
+        return 1, width
+    return width, 1
+
+
+class WorkerConfigMetadata(BaseModel):
```
Comment (simone-chen) [inline scripts/fpm_accuracy/types/worker_config.py]: [minor] Remove unused presentation metadata

`WorkerConfigMetadata` and its private conversion helpers have no callers in this port. The public exporter reads configuration fields directly. Remove this class and its helpers (roughly 280 lines), keeping the worker schema and MoE mapping used by the evaluator.

**Review by simone-chen — COMMENTED** [review]
- Self-review originally covered `cd1844ba`; rechecked the update at `649f9984`.
- [major] The pinned huggingface-hub 1.32 shared blob cache caused the real campaign to fail before evaluation: https://github.com/ai-dynamo/aisimulate/actions/runs/35275817792. Commit `649f9984` now addresses this with a marked shared-store boundary and regression coverage. Full campaign success still needs confirmation.
- [minor] The unused presentation metadata below can be removed without changing the evaluator.
- Validation on the original reviewed head: 317 Python tests, 56 JavaScript tests, and Chromium fixture checks passed; these did not establish real HF campaign success.

File: scripts/fpm_accuracy/types/worker_config.py (hunk 1 above)
Comment (simone-chen) [reply scripts/fpm_accuracy/types/worker_config.py]: - Fixed in `401007cd`: removed the unused metadata class, private converters, and stale docstring references; updated the evaluation README.
- Verified both this cleanup and the existing HF shared-cache fix in `649f9984`: 91 focused FPM tests passed, including shared-store acceptance and boundary rejection. Ruff lint/format and `git diff --check` passed.
- The cleanup is pushed to this PR with DCO sign-off. Full hosted campaign completion remains separate from these local checks.

File: scripts/fpm_accuracy/hf/dataset.py [hunk 2]
```diff
@@ -0,0 +1,1146 @@
...
+def _is_relative_to(path: Path, root: Path) -> bool:
+    try:
+        path.relative_to(root)
+    except ValueError:
+        return False
+    return True
+
+
+def _worker_config_record(configuration_id: str, manifest: Mapping[str, Any]) -> WorkerConfigRecord:
+    model_kind = "moe" if int(manifest.get("moe_ep", 1)) > 1 or int(manifest.get("moe_tp", 1)) > 1 else None
```
[the remark this answers is not in the document]
Comment (simone-chen) [reply scripts/fpm_accuracy/hf/dataset.py]: - Fixed in `4fb17aa0`: validate both MoE dimensions, default absent values to one, and reuse the normalized values for classification and FPM identity.
- Tests cover missing dimensions and null, string, boolean, fractional, zero, and negative values.

File: scripts/fpm_accuracy/hf/protocols.py [hunk 3]
```diff
@@ -0,0 +1,1299 @@
...
+        corrected_times.append(received_s + offsets_s[rank.dp_rank])
+    if corrected_times:
+        if min(corrected_times) != window_start_s:
+            raise DataError(
+                f"measurement file {file.path} row {source_row} grouping.window_start is not the earliest "
+                "corrected receive time"
+            )
+        if any(time < window_start_s or time - window_start_s >= window_size_s for time in corrected_times):
+            raise DataError(f"measurement file {file.path} row {source_row} contains a rank outside grouping.window")
+    return window_start_text, (float(window_start_s),)
```
[the remark this answers is not in the document]
Comment (simone-chen) [reply scripts/fpm_accuracy/hf/protocols.py]: - Fixed in `4fb17aa0`: listener keys now use milliseconds and the same two-part key shape as single-rank events.
- A mixed canonical-stream test verifies that an earlier single-rank event sorts before the listener window and a later event sorts after it.

File: scripts/fpm_accuracy/models/aic_config.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        "nvidia_h200_141gb": "h200_sxm",
+        "nvidia_h200_sxm_141gb": "h200_sxm",
+        "nvidia_b200": "b200_sxm",
+    }.get(normalized, normalized)
+
+
+def _normalize_dtype(value: Any) -> str | None:
+    if value is None:
+        return None
+    return _DTYPE_ALIASES.get(str(value).strip().lower())
```
[the remark this answers is not in the document]
Comment (simone-chen) [reply scripts/fpm_accuracy/models/aic_config.py]: - Fixed in `4fb17aa0`: unknown precision values raise `ConfigurationError` through worker-schema, embedded-config, and override entry points.
- Preserve the existing AISim KV-only mappings for `fp8_e4m3` and `fp8_e5m2`; these are rejected as weight aliases. All 20 configurations at the pinned HF revision pass mapping validation.

File: scripts/fpm_accuracy/models/worker_regression.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        key = (metadata["worker_id"], metadata["regression_bucket"])
+        prior = self._observations.get(key, 0)
+        metadata.update(regression_prior_observations=prior, regression_warmup=prior < self.min_observations)
+        return metadata
+
+    def predict(self, features: ForwardPassInput) -> Prediction:
+        metadata = self.routing_metadata(features)
+        child = self._children[metadata["worker_id"]]
+        stores = child.diagnostics()["regression_stores"]
+        selected = next(store for store in stores if store["workload_kind"] == metadata["regression_bucket"])
```
[the remark this answers is not in the document]
Comment (simone-chen) [reply scripts/fpm_accuracy/models/worker_regression.py:145]: - Fixed in `4fb17aa0`: validate the expected regression stores during child construction. Missing and available bucket names reach the Actions construction-error log; partially created children are closed.
- Tests verify diagnostics, cleanup, and coverage retention. Store validation also passed against the installed native wheel for prefill, decode, and aggregated roles.

File: scripts/fpm_accuracy/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,5 @@
+# Evaluation dependencies; the exact AISim wheel is installed separately.
+huggingface-hub==1.32.0
+pyarrow==25.0.1
+pydantic==2.13.5
+PyYAML==6.0.3
```
[the remark this answers is not in the document]
Comment (simone-chen) [reply scripts/fpm_accuracy/requirements.txt]: - Added `requirements.in` and a complete hash-pinned evaluator dependency graph in `4fb17aa0`; the workflow installs it with `--require-hashes`. Linux wheel downloads passed hash verification, and a clean installation passed `pip check`.
- AISim runtime dependencies remain selected by each exact branch wheel, because main and releases may declare different requirements. This boundary is documented; the evaluator lock does not claim to freeze every evaluated wheel's runtime graph.

File: scripts/prepare_e2e_accuracy_pages.py [hunk 7]
```diff
@@ -463,19 +463,18 @@ def prepare(repo: Path, output: Path) -> None:
...
+            result = subprocess.run(
+                ["git", "-C", str(repo), "show", "origin/" + branch + ":" + path],
+                capture_output=True,
+                text=True,
+            )
+            if result.returncode == 0:
+                committed = result.stdout
+                break
+        if committed is not None:
+            previous = _accuracy_summary(committed)["snapshot"]
```
[the remark this answers is not in the document]
Comment (simone-chen) [reply scripts/prepare_e2e_accuracy_pages.py]: - Fixed in `4fb17aa0`: reuse `committed_accuracy` for both current and legacy snapshot paths. Git failures now propagate instead of bypassing previous-snapshot checks.
- A real temporary Git repository test distinguishes an absent summary path from an unreadable branch ref. Existing publication timestamp and legacy-path tests pass.

File: scripts/prepare_fpm_accuracy_pages.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+                if number not in attempts:
+                    endpoint = f"actions/runs/{run['id']}/attempts/{number}"
+                    attempt = run if number == run["run_attempt"] else api(endpoint)
+                    require(attempt["id"] == run["id"] and attempt["head_sha"] == run["head_sha"], "wrong campaign")
+                    attempts[number] = attempt, api_items(endpoint + "/jobs", "jobs")
+                attempt, jobs = attempts[number]
+                summary = validate_artifact(archive, attempt, artifact["name"], jobs)
+            except (ValueError, KeyError, TypeError, zipfile.BadZipFile) as exc:
+                print(f"Ignoring invalid FPM artifact {artifact['id']}: {exc}")
+                continue
```
[the remark this answers is not in the document]
Comment (simone-chen) [reply scripts/prepare_fpm_accuracy_pages.py:133]: - Fixed in `4fb17aa0`: skip HTTP 404/410 artifacts and continue selecting prior valid results. Other HTTP failures still propagate.
- Tests cover deletion/expiration races, preservation of prior results, and fatal 403/500 responses.

File: tests/fpm_accuracy/test_hf_dataset.py [hunk 9]
```diff
@@ -0,0 +1,1387 @@
...
+        f"version: 1\noverrides:\n  - configuration_path: {CONFIGURATION_PATH}\n",
+        encoding="utf-8",
+    )
+    with pytest.raises(ConfigurationError, match="at least one binding or correction"):
+        _build_dataset(
+            tmp_path / "dataset",
+            protocol_id="forward-pass-measurement-v1",
+            files=[],
+        )
+        HfDataset.from_local(tmp_path / "dataset", revision=REVISION, overrides_path=override)
```
[the remark this answers is not in the document]
Comment (simone-chen) [reply tests/fpm_accuracy/test_hf_dataset.py:1427]: - Fixed in `4fb17aa0`: dataset construction now occurs before `pytest.raises`; only the override load is inside the assertion. The test passes.

File: tests/fpm_accuracy/test_publication.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+        head_sha=summary["snapshot"]["evaluator_sha"],
+        head_branch="main",
+        path=publish.WORKFLOW,
+        status="completed",
+        conclusion="failure",
+        event="schedule",
+        repository={"full_name": publish.REPO},
+        head_repository={"full_name": publish.REPO},
+        **updates,
+    )
```
[the remark this answers is not in the document]
Comment (simone-chen) [reply tests/fpm_accuracy/test_publication.py:51]: - Added in `4fb17aa0`: independent rejection tests for foreign source/producer repositories, the wrong workflow/event/branch, incomplete runs, and cancelled runs. Each reaches the FPM untrusted-campaign gate.

File: tests/fpm_accuracy/test_publication.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+    assert workflow["jobs"]["branches"]["strategy"]["max-parallel"] == "2"
+    assert "pull_request" not in workflow["on"]
+    branch = yaml.load((ROOT / ".github/workflows/fpm-accuracy-branch.yml").read_text(), Loader=yaml.BaseLoader)
+    campaign = branch["jobs"]["campaign"]
+    assert campaign["needs"] == "wheel"
+    upload = next(s for s in campaign["steps"] if "upload-artifact@" in s.get("uses", ""))
+    assert upload["with"]["retention-days"] == "90" and "if" not in upload
+    # Container hooks must remap one directory, not multiline host paths.
+    assert upload["with"]["path"] == "${{ runner.temp }}/accuracy-public/"
+    assert "gitlab" not in json.dumps(workflow) + json.dumps(branch)
```
[the remark this answers is not in the document]
Comment (simone-chen) [reply tests/fpm_accuracy/test_publication.py]: - Fixed in `4fb17aa0`: lowercase the serialized workflow data before checking for the forbidden fragment. The workflow contract test passes.

File: tests/test_fpm_accuracy_workflow.mjs [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+    listWorkflowRuns: async (args) => {
+      assert.equal(args.head_sha, sha);
+      assert.equal(args.event, "schedule");
+      return { data: { workflow_runs: runs } };
+    },
+    listJobsForWorkflowRun: "jobs",
+    listWorkflowRunArtifacts: "artifacts",
+  };
+  const context = vm.createContext({
+    fetch: async () => ({ok: true, json: async () => ({sha: "d".repeat(40)})}),
```
[the remark this answers is not in the document]
Comment (simone-chen) [reply tests/test_fpm_accuracy_workflow.mjs]: - Added in `4fb17aa0`: configurable HF responses and rejection cases for HTTP failure, empty revision, branch text, a 39-character SHA, and a 40-character non-hex value. JavaScript workflow tests pass.

File: tests/test_fpm_accuracy_workflow.mjs [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+        const github = {
+          event_name: "workflow_run", ref: "refs/heads/main", repository: "ai-dynamo/aisimulate",
+          event: { workflow_run: { name, conclusion, head_repository: { full_name: repository } } },
+        };
+        assert.equal(vm.runInNewContext(expression, { github }), repository === github.repository &&
+          (conclusion === "success" || (["E2E Accuracy Matrix", "FPM Accuracy Matrix"].includes(name) && conclusion === "failure")));
+      }
+    }
+  }
+});
```
[the remark this answers is not in the document]
Comment (simone-chen) [reply tests/test_fpm_accuracy_workflow.mjs]: - Removed the duplicate FPM test in `4fb17aa0`. The E2E workflow test retains the same Pages truth-table coverage, including failed FPM campaigns. Both workflow suites pass.

File: scripts/fpm_accuracy/hf/dataset.py [hunk 14]
```diff
@@ -0,0 +1,1148 @@
...
+                )
+            )
+        return tuple(result)
+
+    def _read_json(self, relative: str, *, expected_sha256: str | None = None) -> Mapping[str, Any]:
+        path = self._safe_path(relative)
+        if expected_sha256 is not None:
+            self._verify_file(path, expected_sha256, description=relative)
+        try:
+            value = json.loads(path.read_text(encoding="utf-8"))
```
Comment (jasonqinzhou) [inline scripts/fpm_accuracy/hf/dataset.py]: Please parse the HF catalog, manifests, and sidecars with duplicate-key rejection. `json.loads` silently keeps the last occurrence, so an input containing two values for an identity or provenance field is accepted even though the bytes are ambiguous; pinning and hashing that file do not resolve which value the producer intended. The repository review contract requires corrupt or ambiguous product evidence to fail loudly. A strict `object_pairs_hook` here, shared by the other dataset JSON readers, plus a negative test would close the gap.

File: pages/fpm-accuracy/app.js [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+  }
+
+  function render() {
+    const models = groups();
+    body.innerHTML = models.map((group) => {
+      const expanded = !collapsed.has(group.model);
+      const model = `<tr class="overview-model-row model-row"><th scope="rowgroup"><button class="overview-model-button" data-model="${escape(group.model)}" aria-expanded="${expanded}"><span class="overview-chevron" aria-hidden="true">›</span><span>${escape(group.model)}</span><span class="overview-model-count">${group.rows.length}</span></button></th><td>${escape(group.gpu)}</td><td>${escape(group.framework)}</td><td class="overview-measurement-cell"><strong>${integer(group.measurements)}</strong><span>observations</span></td>${cells(group.rows)}</tr>`;
+      return model + [...group.rows].sort((a, b) => [a.gpu, a.framework, a.framework_version, a.parallelism].join().localeCompare([b.gpu, b.framework, b.framework_version, b.parallelism].join())).map((row) => {
+        const measurement = row.status === "ready" ? `${integer(row.measurement_count)} observations` : row.status.replaceAll("_", " ");
+        const skipped = row.skipped_count ? `<span>${integer(row.skipped_count)} excluded</span>` : "";
```
Comment (jasonqinzhou) [inline pages/fpm-accuracy/app.js]: `skipped_count` is not an excluded-only count: `evaluate_case` sums every `MeasurementIssue`, while the protocol adapters emit both `EXCLUDED` and `MEASUREMENT_UNAVAILABLE` issues (for example, unsynchronized DP streams and incomplete synchronized iterations). Rendering the combined total as `excluded` turns missing or incomplete measurement evidence into what looks like deliberate filtering. Please preserve separate unavailable/excluded counts in the public contract, or use a neutral label that does not erase that distinction.

**Review by jasonqinzhou — APPROVED** [review]
**APPROVE · 8/10 · high confidence**

## What this PR does

This PR adds a daily, exact-revision FPM accuracy campaign for main and supported release branches, publishes qualified aggregates through GitHub Pages, and moves the public static site into a root pages directory. It also ports the dataset adapters and predictor harness needed to compare FPM with KV warmup on and off against worker-isolated regression and pinned Hugging Face measurements.

The update responds carefully to review feedback by normalizing mixed-stream chronology, validating MoE and precision metadata, checking regression-store contracts, preserving container artifacts, hash-locking campaign dependencies, and adding focused negative tests around each boundary.

## Why this score

The current head is generally sound and the incremental fixes pass broad focused verification, so the score is 8.0/10 with a COMMENT recommendation. Two non-blocking P2 evidence-quality gaps remain: duplicate keys in pinned dataset JSON are still accepted as last-value-wins input, and the public overview still labels unavailable source observations as excluded. The requested Fable model was not observed, so this result retains the independent Codex recommendation but is not dual-reviewed.

File: scripts/fpm_accuracy/hf/dataset.py (hunk 14 above)
Comment (simone-chen) [reply scripts/fpm_accuracy/hf/dataset.py]: Fixed in 90c08444943ad2881884654fe6b632ac61673634. Catalog, configuration/measurement manifest, and FPM sidecar readers now reuse contract.strict_json, including the native staging sidecar reader. Five negative cases cover top-level and nested duplicates; the measurement-manifest hash is updated to ensure hash validation cannot mask the parsing check. All 368 focused Python tests passed.

File: pages/fpm-accuracy/app.js (hunk 15 above)
Comment (simone-chen) [reply pages/fpm-accuracy/app.js]: Fixed in 90c08444943ad2881884654fe6b632ac61673634. The combined count now reads “excluded or unavailable,” preserving the existing artifact schema without implying deliberate filtering. Documented the count semantics and added a Chromium assertion for a nonzero combined count; browser checks passed.

**Review by simone-chen — COMMENTED** [review]
- Codex follow-up review of 90c08444943ad2881884654fe6b632ac61673634: traced strict JSON parsing through catalog, configuration/measurement manifests, hash-verified FPM sidecars, and native staging; checked the combined unavailable/excluded count through the public renderer. No remaining finding in this update.
- Both new review findings are fixed, tested, and resolved. The main merge only required generated CODEOWNERS reconciliation; the resulting ownership changes match the declared FPM/Pages sources.
- Validation: 368 focused Python tests, 21 workflow JavaScript tests, Chromium fixture checks, Ruff, packaged legal notices, and generated CODEOWNERS checks passed. Hosted checks and the required Full CI are still pending; local tests are not a claim of full merge readiness.

**Comment by simone-chen** [comment]
- Verified merged commit 4acab657e054b0ac135599b3a5206899cf831c3e includes both final review fixes; all review threads are resolved.
- Fast CI, CodeRabbit, Pages build, CODEOWNERS, and DCO passed on reviewed head 90c08444943ad2881884654fe6b632ac61673634. The main Pages deployment also succeeded: https://github.com/ai-dynamo/aisimulate/actions/runs/35294682332.
- Full CI is not a passing result: https://github.com/ai-dynamo/aisimulate/actions/runs/35294618973. The baseline prediction-snapshot job failed at `git show ${GITHUB_SHA}:scripts/ci_install_build_tools.sh` because that helper is absent from the reviewed commit. The PR had already been merged independently while this run was in progress; the remaining jobs were canceled after the failure was diagnosed.
